### PR TITLE
Addition of Databox Service Module to Azure Powershell

### PR DESCRIPTION
**Pull request by t-irali**
## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (Microsoft internal partners only). Link to the design review : https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/328
       
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by adxsdkps** [comment]
Can one of the admins verify this patch?

**Comment by t-irali** [comment]
## Issues
- Some checks are failing stating "Online version in the header of the file is incorrect." in the help files of the cmdlets. I dont have an online documentation yet. How do we solve this?

- In the testing guidelines, it is mentioned that the tests must not have any hard coded values. But some cmdlets in the module require address as parameters, so I had to pass a hard coded address. I dint find any other way to do it.

File: src/DataBox/DataBox.Test/SessionRecords/desktop.ini [hunk 1]
```diff
@@ -0,0 +1,4 @@
+[ViewState]
```
Comment (markcowl) [inline src/DataBox/DataBox.Test/SessionRecords/desktop.ini]: Is this used anywhere?  If not, please remove

File: src/DataBox/DataBox/Az.DataBox.psd1 [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+# Generated on: 4/18/2019
+#
+
+@{
+
+# Script module or binary module file associated with this manifest.
+# RootModule = ''
+
+# Version number of this module.
+ModuleVersion = '1.0.0'
```
Comment (markcowl) [inline src/DataBox/DataBox/Az.DataBox.psd1]: 0.1.0

File: src/DataBox/DataBox/Az.DataBox.psd1 [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        LicenseUri = 'https://aka.ms/azps-license'
+
+        # A URL to the main website for this project.
+        ProjectUri = 'https://github.com/Azure/azure-powershell'
+
+        # A URL to an icon representing this module.
+        # IconUri = ''
+
+        # ReleaseNotes of this module
+        ReleaseNotes = '* Updated cmdlets with plural nouns to singular, and deprecated plural names.'
```
Comment (markcowl) [inline src/DataBox/DataBox/Az.DataBox.psd1]: Initial release of Azure DataBox cmdlets

File: src/DataBox/DataBox/Common/Cmdlets/GetDataBoxJobCmdlet.cs [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+        public override void ExecuteCmdlet()
+        {
+            if (this.ParameterSetName.Equals("GetByResourceIdParameterSet"))
+            {
+                this.ResourceGroupName = ResourceIdHandler.GetResourceGroupName(ResourceId);
+                this.Name = ResourceIdHandler.GetResourceName(ResourceId);
+            }
+
+            if (Name != null && string.IsNullOrWhiteSpace(Name))
```
Comment (markcowl) [inline src/DataBox/DataBox/Common/Cmdlets/GetDataBoxJobCmdlet.cs]: You don't need this check, the validation attribute will do this for you

File: src/DataBox/DataBox/Common/Cmdlets/GetDataBoxJobCmdlet.cs [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+                 IPage<JobResource> jobPageList = null;
+                 List<JobResource> result = new List<JobResource>();
+                 List<PSDataBoxJob> finalResult = new List<PSDataBoxJob>();
+
+
+                do
+                {
+                     // Lists all the jobs available under the subscription.
+                     if (jobPageList == null)
```
Comment (markcowl) [inline src/DataBox/DataBox/Common/Cmdlets/GetDataBoxJobCmdlet.cs]: seems like you can use a single method to expland pages and filter the list, if you just pass in a function to get the job page list

File: src/DataBox/DataBox/Common/Cmdlets/GetDataBoxJobCmdlet.cs [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+                                || (CompletedWithErrors && job.Status == StageName.CompletedWithErrors))
+                            {
+                                result.Add(job);
+                            }
+                        }
+                    }
+                    else
+                    {
+
+                        result.AddRange(jobPageList.ToList());
```
Comment (markcowl) [inline src/DataBox/DataBox/Common/Cmdlets/GetDataBoxJobCmdlet.cs]: Not that you can write these to the pipeline as you discover them.

File: src/DataBox/DataBox/Common/Cmdlets/NewDataBoxJobCmdlet.cs [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+
+            if (addressValidationResult.ValidationStatus != AddressValidationStatus.Valid)
+            {
+                Console.WriteLine("Address validation status: {0}", addressValidationResult.ValidationStatus);
+
+                //print alternate address
+                if (addressValidationResult.ValidationStatus == AddressValidationStatus.Ambiguous)
+                {
+                    Console.WriteLine("\nSUPPORT ADDRESSES: ");
```
Comment (markcowl) [inline src/DataBox/DataBox/Common/Cmdlets/NewDataBoxJobCmdlet.cs]: Do not use Console.WriteLine in cmdlet code.  If this is meant as a message, use WriteVerbose.  If this is meant to be the ouptut of the cmdlet, you should return an object and you can alter the output format to write the data the way you want it to look.

File: src/DataBox/DataBox/Common/Cmdlets/NewDataBoxJobCmdlet.cs [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+                            Console.WriteLine("State/Province: {0}", address.StateOrProvince);
+                        if (!(string.IsNullOrEmpty(address.Country)))
+                            Console.WriteLine("Country: {0}", address.Country);
+                        if (!(string.IsNullOrEmpty(address.PostalCode)))
+                            Console.WriteLine("Postal code: {0}", address.PostalCode);
+                        if (!(string.IsNullOrEmpty(address.ZipExtendedCode)))
+                            Console.WriteLine("Zip extended code: {0}", address.ZipExtendedCode);
+                        Console.WriteLine("\n\n");
+                    }
+                    throw new PSNotSupportedException("\nThe Shipping Address is ambiguous. Please select any address from the ones provided above.");
```
Comment (markcowl) [inline src/DataBox/DataBox/Common/Cmdlets/NewDataBoxJobCmdlet.cs]: Please use resource strings for customer-facing text

File: src/DataBox/DataBox/Common/Cmdlets/NewDataBoxJobCmdlet.cs [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+                }
+                throw new PSNotSupportedException("The Shipping Address is not Valid.");
+            }
+
+            if (ShouldProcess(this.Name))
```
Comment (markcowl) [inline src/DataBox/DataBox/Common/Cmdlets/NewDataBoxJobCmdlet.cs]: use target and action strings with ShouldProcess

File: src/DataBox/DataBox/Common/Cmdlets/NewDataBoxJobCmdlet.cs [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+            {
+                JobResource finalJobResource = JobsOperationsExtensions.Create(
+                            DataBoxManagementClient.Jobs,
+                            ResourceGroupName,
+                            Name,
+                            newJobResource);
+
+                WriteObject(new PSDataBoxJob(finalJobResource));
+            }
+
```
Comment (markcowl) [inline src/DataBox/DataBox/Common/Cmdlets/NewDataBoxJobCmdlet.cs]: remove unneeded vertical space

File: src/DataBox/DataBox/Common/Cmdlets/RemoveDataBoxJobCmdlet.cs [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+        [Parameter(Mandatory = true, ParameterSetName = GetByNameParameterSet)]
+        [ValidateNotNullOrEmpty]
+        public string Name { get; set; }
+
+        [Parameter(Mandatory = true, ParameterSetName = GetByResourceIdParameterSet, ValueFromPipelineByPropertyName = true)]
+        [Alias("Id")]
+        [ValidateNotNullOrEmpty]
+        public string ResourceId { get; set; }
+
```
Comment (markcowl) [inline src/DataBox/DataBox/Common/Cmdlets/RemoveDataBoxJobCmdlet.cs]: Need InputObject parameter set

File: src/DataBox/DataBox/Common/Cmdlets/RemoveDataBoxJobCmdlet.cs [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+                this.ResourceGroupName = ResourceIdHandler.GetResourceGroupName(ResourceId);
+                this.Name = ResourceIdHandler.GetResourceName(ResourceId);
+            }
+            // Gets information about the specified job.
+            JobResource jobResource = JobsOperationsExtensions.Get(DataBoxManagementClient.Jobs, ResourceGroupName, Name, "details");
+
+            if (jobResource.Status == StageName.Cancelled || jobResource.Status == StageName.Completed || jobResource.Status == StageName.CompletedWithErrors)
+            {
+                // Initiate to delete job
+                if (ShouldProcess(this.Name))
```
Comment (markcowl) [inline src/DataBox/DataBox/Common/Cmdlets/RemoveDataBoxJobCmdlet.cs]: use target and action strings in ShouldProcess

File: src/DataBox/DataBox/Common/Cmdlets/RemoveDataBoxJobCmdlet.cs [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+                {
+                    JobsOperationsExtensions.Delete(
+                        DataBoxManagementClient.Jobs,
+                        ResourceGroupName,
+                        Name);
+
+                    WriteObject("Successfully Deleted the Databox Job");
```
Comment (markcowl) [inline src/DataBox/DataBox/Common/Cmdlets/RemoveDataBoxJobCmdlet.cs]: In Remove cmdlets
- return nothing on successful deletion
- Add a -PassThru swtich parameter, and if this is present, write out the single bool value `true`

File: src/DataBox/DataBox/Common/Cmdlets/StopDataBoxJobCmdlet.cs [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+        [Parameter(Mandatory = true, HelpMessage = "Reason For Cancellation")]
+        [ValidateNotNullOrEmpty]
+        public string Reason { get; set; }
+
+        [Parameter(Mandatory = true, ParameterSetName = GetByResourceIdParameterSet,  ValueFromPipelineByPropertyName = true)]
+        [Alias("Id")]
+        [ValidateNotNullOrEmpty]
+        public string ResourceId { get; set; }
+
```
Comment (markcowl) [inline src/DataBox/DataBox/Common/Cmdlets/StopDataBoxJobCmdlet.cs]: Needs an InputObject parameter set, taking a PSDataBoxJob

File: src/DataBox/DataBox/Common/Cmdlets/StopDataBoxJobCmdlet.cs [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+                        Name,
+                        Reason);
+                }
+                
+               
+                WriteObject("Successfully Cancelled the Databox Job");
+            }
+            else if(jobResource.Status == StageName.Cancelled)
+            {
+                WriteObject("This Databox Job is already Cancelled");
```
Comment (markcowl) [inline src/DataBox/DataBox/Common/Cmdlets/StopDataBoxJobCmdlet.cs]: As with Remove, this should return nothing on successful completion, and should provide a PassThru parameter that return 'true' is present

File: src/DataBox/DataBox/Common/ResourceIDHandler.cs [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+        {
+            var splits = resourceId.Split(new[] { '/' });
+            for(int i = 0;i<splits.Length; i++)
+            {
+                if (splits[i].Equals("resourceGroups", StringComparison.CurrentCultureIgnoreCase))
+                {
+                    return splits[i + 1];
+                }
+            }
+            throw new Exception("Invalid Resource Id");
```
Comment (markcowl) [inline src/DataBox/DataBox/Common/ResourceIDHandler.cs]: Put customer-facing strings into project resources

File: src/DataBox/DataBox/Common/ResourceIDHandler.cs [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+        }
+
+        public static string GetResourceName(string resourceId)
+        {
+            var splits = resourceId.Split(new[] { '/' });
+            for (int i = 0; i < splits.Length; i++)
+            {
+                if (splits[i].Equals("providers", StringComparison.CurrentCultureIgnoreCase))
```
Comment (markcowl) [inline src/DataBox/DataBox/Common/ResourceIDHandler.cs]: same comment on customer-facing strings

File: src/DataBox/DataBox/DataBox.csproj [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+  <Import Project="$(MSBuildThisFileDirectory)..\..\Az.props" />
+
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Azure.Management.DataBox" Version="1.0.0" />
+    <PackageReference Include="Microsoft.Azure.Management.Storage" Version="12.0.0" />
+  </ItemGroup>
+
+
+  <ItemGroup>
```
Comment (markcowl) [inline src/DataBox/DataBox/DataBox.csproj]: You may not use this dependency

File: src/DataBox/DataBox/DataBox.csproj [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+    <AssemblyName>Microsoft.Azure.PowerShell.Cmdlets.DataBox</AssemblyName>
+    <RootNamespace>Microsoft.Azure.PowerShell.Cmdlets.DataBox</RootNamespace>
+  </PropertyGroup>
+
+  <Import Project="$(MSBuildThisFileDirectory)..\..\Az.props" />
+
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Azure.Management.DataBox" Version="1.0.0" />
+    <PackageReference Include="Microsoft.Azure.Management.Storage" Version="12.0.0" />
```
Comment (markcowl) [inline src/DataBox/DataBox/DataBox.csproj]: You cannot take a dependency on storage management dll - this conflicts with the dependency in Az.Storage.  Instead, use the common storage management client in this package that you alreadyu have a dependency on: https://github.com/Azure/azure-powershell/blob/master/tools/Common.Netcore.Dependencies.targets#L18

File: src/DataBox/DataBox/Help/Az.DataBox.md [hunk 20]
```diff
@@ -0,0 +1,28 @@
+---
+Module Name: Az.DataBox
+Module Guid: '39B6B3E6-A15E-48BA-BE20-FE0D441D92B4'
+Download Help Link: {{ Update Download Link }}
```
Comment (markcowl) [inline src/DataBox/DataBox/Help/Az.DataBox.md]: You will need to fill these out

File: src/DataBox/DataBox/Help/Az.DataBox.md [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+Help Version: {{ Update Help Version }}
+Locale: {{ Update Locale }}
+---
+
+# Az.DataBox Module
+## Description
+Module to Create, Cancel, Delete and Fetch Databox Jobs
+
+## Az.DataBox Cmdlets
+### [Get-AzDataBoxJob](Get-AzDataBoxCredential.md)
```
Comment (markcowl) [inline src/DataBox/DataBox/Help/Az.DataBox.md]: bad link

File: src/DataBox/DataBox/Help/Get-AzDataBoxCredential.md [hunk 22]
```diff
@@ -0,0 +1,132 @@
+---
+external help file: Microsoft.Azure.PowerShell.Cmdlets.DataBox.dll-Help.xml
+Module Name: Az.DataBox
+online version:
```
Comment (markcowl) [inline src/DataBox/DataBox/Help/Get-AzDataBoxCredential.md]: You will need to fill this out for each of these files.  https://docs.microsoft.com/en-us/powershell/module/az.databox/get-azdataboxcredential

File: src/DataBox/DataBox.Test/ScenarioTests/DataBoxTests.cs [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+
+        public DataBoxTests(Xunit.Abstractions.ITestOutputHelper output)
+        {
+            _logger = new XunitTracingInterceptor(output);
+            XunitTracingInterceptor.AddToContext(_logger);
+        }
+
+
+#if NETSTANDARD
+        [Fact]
```
[the remark this answers is not in the document]
Comment (t-irali) [reply src/DataBox/DataBox.Test/ScenarioTests/DataBoxTests.cs]: I used the testing template from the data factory service solution. No specific reason to include them. I removed it.

File: src/DataBox/DataBox/Common/Cmdlets/GetDataBoxJobCmdlet.cs (hunk 5 above)
Comment (t-irali) [reply src/DataBox/DataBox/Common/Cmdlets/GetDataBoxJobCmdlet.cs]: I did not understand. Anyways is this not okay to do?

File: src/DataBox/DataBox/Common/Cmdlets/GetDataBoxJobCmdlet.cs (hunk 6 above)
Comment (t-irali) [reply src/DataBox/DataBox/Common/Cmdlets/GetDataBoxJobCmdlet.cs]: I want to write it once to the pipeline. Is it not okay?

File: src/DataBox/DataBox/DataBox.csproj (hunk 18 above)
Comment (t-irali) [reply src/DataBox/DataBox/DataBox.csproj]: Which dependency are you talking about?

**Comment by t-irali** [comment]
The 3 checks (windows-powershell, windows-powershell (analyze) and windows-powershell (test) ) are failing because the current Databox Management dll has sha1 signature. We have started the process of publishing the dll with sha256 signature. It might take a couple of days. After that the issue must be gone

File: src/DataBox/documentation/current-breaking-changes.md [hunk 24]
```diff
@@ -0,0 +1,41 @@
+<!--
```
Comment (markcowl) [inline src/DataBox/documentation/current-breaking-changes.md]: Remove the 'Microsoft.Azure.Management.Databox.dll' file above.

**Comment by t-irali** [comment]
@markcowl  can you please let me know how do we put up an online help documentation on https://docs.microsoft.com/en-us/powershell this link?

**Comment by markcowl** [comment]
@t-irali online help ships with the module.  Documentation is only updated at release milestones (next update 7/30)

**Pull request merged**